### PR TITLE
optimize consensus

### PR DESCRIPTION
**Pull request by erikzhang**
Set transactions and send hashes on InitializeConsensus

### CODE REVIEW

File: neo/Consensus/ConsensusService.cs [hunk 1]
```diff
@@ -155,6 +171,16 @@ private void InitializeConsensus(byte view_number)
                 if (context.MyIndex == context.PrimaryIndex)
                 {
                     context.State |= ConsensusState.Primary;
+                    if (!context.State.HasFlag(ConsensusState.SignatureSent))
```
Comment (shargon) [inline neo/Consensus/ConsensusService.cs]: Is possible that this state returns `false`?

File: neo/Consensus/ConsensusService.cs [hunk 2]
```diff
@@ -134,6 +115,41 @@ public void Dispose()
...
+                Value = amount_netfee,
+                ScriptHash = wallet.GetChangeAddress()
+            } };
+            while (true)
+            {
+                ulong nonce = GetNonce();
+                MinerTransaction tx = new MinerTransaction
+                {
+                    Nonce = (uint)(nonce % (uint.MaxValue + 1ul)),
+                    Attributes = new TransactionAttribute[0],
```
Comment (shargon) [inline neo/Consensus/ConsensusService.cs]: Maybe we can force the unique hash adding the block index

`                Attributes = new TransactionAttribute[] { new TransactionAttribute() { Data = BitConverter.GetBytes(context.BlockIndex), Usage = TransactionAttributeUsage.Remark } },
`

File: neo/Consensus/ConsensusService.cs (hunk 1 above)
Comment (erikzhang) [reply neo/Consensus/ConsensusService.cs]: Yes. This will happen after a `changeview` sometimes.

File: neo/Consensus/ConsensusService.cs (hunk 2 above)
Comment (erikzhang) [reply neo/Consensus/ConsensusService.cs]: Why not just using `Nonce`?

**Comment by erikzhang** [comment]
@shargon Can I merge it now?

File: neo/Consensus/ConsensusService.cs (hunk 1 above)
Comment (shargon) [reply neo/Consensus/ConsensusService.cs]: But if this is false, how they Fill the TXs?

Comment (erikzhang) [reply neo/Consensus/ConsensusService.cs]: They needn't. The transactions have been filled  before changing view.

**Pull request merged**